### PR TITLE
TPS Warp and MNE_BEM fixes

### PR DESCRIPTION
**Pull request by JanaKiesel**

### CODE REVIEW

File: MNE/mne/mne_bem.cpp [hunk 1]
```diff
@@ -123,8 +117,7 @@ bool MNEBem::readFromStream(FiffStream::SPtr& p_pStream, bool add_geom, FiffDirT
         QList<FiffDirEntry> t_Dir;
         QString t_sFileName = p_pStream->streamName();
 
-        t_file.setFileName(t_sFileName);
-        p_pStream = FiffStream::SPtr(new FiffStream(&t_file));
+        t_file.setFileName(t_sFileName);        p_pStream = FiffStream::SPtr(new FiffStream(&t_file));
```
Comment (LorenzE) [inline MNE/mne/mne_bem.cpp]: add linebreak

File: MNE/mne/mne_bem.cpp [hunk 2]
```diff
@@ -161,17 +159,19 @@ bool MNEBem::readFromStream(FiffStream::SPtr& p_pStream, bool add_geom, FiffDirT
...
-
 //           src(k) = this;
     }
 
+    printf("\t%d bem surfaces read\n", bemsurf.size());
+
     if(open_here)
         p_pStream->device()->close();
-
+//    bool testStream =p_pStream->device()->isOpen();
```
Comment (LorenzE) [inline MNE/mne/mne_bem.cpp]: delete this line

File: MNE/mne/mne_bem.cpp [hunk 3]
```diff
@@ -252,7 +252,8 @@ bool MNEBem::readBemSurface(FiffStream *p_pStream, const FiffDirTree &p_Tree, MN
 
     p_BemSurface.rr = t_pTag->toFloatMatrix().transpose();
     qint32 rows_rr = p_BemSurface.rr.rows();
-        qDebug() << "last element rr: " << p_BemSurface.rr(rows_rr-1, 0) << p_BemSurface.rr(rows_rr-1, 1) << p_BemSurface.rr(rows_rr-1, 2);
+    std::cout << "Here are the first Rows rr:" << std::endl << p_BemSurface.rr.topRows(9) << std::endl;
```
Comment (LorenzE) [inline MNE/mne/mne_bem.cpp]: Delete this line

File: MNE/mne/mne_bem.cpp [hunk 4]
```diff
@@ -252,7 +252,8 @@ bool MNEBem::readBemSurface(FiffStream *p_pStream, const FiffDirTree &p_Tree, MN
...
+//        qDebug() << "last element rr: " << p_BemSurface.rr(rows_rr-1, 0) << p_BemSurface.rr(rows_rr-1, 1) << p_BemSurface.rr(rows_rr-1, 2);
```
Comment (LorenzE) [inline MNE/mne/mne_bem.cpp]: Delete this line

File: MNE/mne/mne_bem.cpp [hunk 5]
```diff
@@ -323,9 +324,90 @@ bool MNEBem::readBemSurface(FiffStream *p_pStream, const FiffDirTree &p_Tree, MN
         MatrixXi p_defaultMatrix(0, 0);
         p_BemSurface.tris = p_defaultMatrix;
     }
+
         qDebug() << "Triangles; type:" << t_pTag->getType() << "rows:" << p_BemSurface.tris.rows() << "cols:" << p_BemSurface.tris.cols();
```
Comment (LorenzE) [inline MNE/mne/mne_bem.cpp]: delete this line and the four following lines too

File: MNE/mne/mne_bem.cpp [hunk 6]
```diff
@@ -323,9 +324,90 @@ bool MNEBem::readBemSurface(FiffStream *p_pStream, const FiffDirTree &p_Tree, MN
...
+{
+    //
+    //   Open the file, create directory
+    //
+
+    // Create the file and save the essentials
+    FiffStream::SPtr t_pStream = FiffStream::start_file(p_IODevice);
+    printf("Write BEM surface in %s...", t_pStream->streamName().toUtf8().constData(), "\n");
+    this->writeToStream(t_pStream.data());
+//    bool test= t_pStream->device()->isOpen();
```
Comment (LorenzE) [inline MNE/mne/mne_bem.cpp]: Delete

File: MNE/mne/mne_bem.cpp [hunk 7]
```diff
@@ -323,9 +324,90 @@ bool MNEBem::readBemSurface(FiffStream *p_pStream, const FiffDirTree &p_Tree, MN
...
+}
+
+//*************************************************************************************************************
+
+void MNEBem::writeToStream(FiffStream* p_pStream)
+{
+    p_pStream->start_block(FIFFB_BEM);
+    for(qint32 h = 0; h < m_qListBemSurface.size(); ++h)
+    {
+        std::cout << "Here are the first Rows rr bevor writing:" << std::endl << m_qListBemSurface[h].rr.topRows(9) << std::endl;
```
Comment (LorenzE) [inline MNE/mne/mne_bem.cpp]: Delete

File: MNE/mne/mne_bem.h [hunk 8]
```diff
@@ -164,11 +164,63 @@ class MNESHARED_EXPORT MNEBem
```
Comment (LorenzE) [inline MNE/mne/mne_bem.h]: Please add the following convenient functions 
void clear();
inline bool isEmpty() const;

See MNESourceSpace for more information

File: MNE/mne/mne_bem_surface.cpp [hunk 9]
```diff
@@ -221,7 +221,26 @@ bool MNEBemSurface::addVertexNormals()
 
  qDebug() << "nn first Row:" << this->nn(0,0)<<this->nn(0,1)<<this->nn(0,2);
```
Comment (LorenzE) [inline MNE/mne/mne_bem_surface.cpp]: Delete this line

File: MNE/utils/warp.cpp [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+* PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+* INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*
+*
+* @brief    Implementation of a thin plate spline warping class.
```
Comment (LorenzE) [inline MNE/utils/warp.cpp]: Rename to "Warp class declaration"

File: MNE/utils/warp.h [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+* WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+* PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+* INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+* PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+* HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+* NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+* POSSIBILITY OF SUCH DAMAGE.
+*
+*
+* @brief    warp class declaration.
```
Comment (LorenzE) [inline MNE/utils/warp.h]: Rename to "Warp class declaration"

File: MNE/utils/warp.h [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+//#include <QString>
+
+//*************************************************************************************************************
+//=============================================================================================================
+// DEFINE NAMESPACE UTILSLIB
+//=============================================================================================================
+
+namespace UTILSLIB
+{
+
```
Comment (LorenzE) [inline MNE/utils/warp.h]: Add linebreak

File: MNE/utils/warp.h [hunk 13]
```diff
@@ -0,0 +1,159 @@
...
+//*************************************************************************************************************
+//=============================================================================================================
+// QT INCLUDES
+//=============================================================================================================
+
+#include <QSharedPointer>
+#include <QDebug>
+#include <QFile>
+//#include <QString>
+
```
Comment (LorenzE) [inline MNE/utils/warp.h]: Add linebreak

File: MNE/utils/warp.h [hunk 14]
```diff
@@ -0,0 +1,159 @@
...
+#include "utils_global.h"
+
+//*************************************************************************************************************
+//=============================================================================================================
+// Eigen INCLUDES
+//=============================================================================================================
+
+#include <Eigen/Core>
+#include <Eigen/LU>
+
```
Comment (LorenzE) [inline MNE/utils/warp.h]: Add linebreak

File: MNE/utils/warp.h [hunk 15]
```diff
@@ -0,0 +1,159 @@
...
+     MatrixXd readsLm(const QString &electrodeFileName);
+
+private:
+    //=========================================================================================================
+    /**
+    * Calculate the weighting parameters.
+    *
+    * @param[in]  sLm      3D Landmarks of the source geometry
+    * @param[in]  dLm      3D Landmarks of the destination geometry
+    * @param[out] warpWeight Weighting parameters of the tps warp
```
Comment (LorenzE) [inline MNE/utils/warp.h]: update docu warpWeight not used anymore?

File: examples/testWarp/main.cpp [hunk 16]
```diff
@@ -0,0 +1,226 @@
+//=============================================================================================================
+/**
+* @file     main.cpp
+* @author   Jana Kiesel <jana.kiesel@tu-ilmenau.de>
+*           Christoph Dinh <chdinh@nmr.mgh.harvard.edu>;
+*           Matti Hamalainen <msh@nmr.mgh.harvard.edu>
+* @version  1.0
+* @date     November, 2015*
```
Comment (LorenzE) [inline examples/testWarp/main.cpp]: remove * behind 2015

File: examples/testWarp/main.cpp [hunk 17]
```diff
@@ -0,0 +1,226 @@
...
+#include <cstdlib>
+#include <time.h>
+
+#include <utils/warp.h>
+#include <fiff/fiff.h>
+#include <mne/mne_bem.h>
+
+
+#include <QFile>
+
```
Comment (LorenzE) [inline examples/testWarp/main.cpp]: Remove linebreak

File: examples/testWarp/main.cpp [hunk 18]
```diff
@@ -0,0 +1,226 @@
...
+//#include <fstream>
+#include <iostream>
+#include <cstdlib>
+#include <time.h>
+
+#include <utils/warp.h>
+#include <fiff/fiff.h>
+#include <mne/mne_bem.h>
+
+
```
Comment (LorenzE) [inline examples/testWarp/main.cpp]: Remove linebreak

File: examples/testWarp/main.cpp [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+
+//*************************************************************************************************************
+//=============================================================================================================
+// QT INCLUDES
+//=============================================================================================================
+
+#include <QtCore/QCoreApplication>
+
```
Comment (LorenzE) [inline examples/testWarp/main.cpp]: Add linebreak

File: MNE/utils/warp.h (hunk 15 above)
Comment (JanaKiesel) [inline MNE/utils/warp.h]: To Do: Implement the outputparameters

File: MNE/mne/mne_bem.h (hunk 8 above)
Comment (JanaKiesel) [inline MNE/mne/mne_bem.h]: To Do

**Comment by LorenzE** [comment]
Fantastic work Jana! :)

**Pull request merged**